### PR TITLE
BUG Prevent brutal discard of constructed Image_Backend instances

### PR DESCRIPTION
**Pull request by tractorcow**
@mikenz thanks for calling us out on our bad coding. :)

With this fix the asset-admin performs a LOT faster.

### CODE REVIEW

**Comment by tractorcow** [comment]
Note: This PR has raised an issue where the InterventionBackend isn't acting in an immutable way. Need to fix this before it can be merged.

**Comment by mikenz** [comment]
Thanks :-) I'll have a play

**Comment by tractorcow** [comment]
It looks as though

```php
$resized = $image->SizeAction(100, 200);
```

Is modifying the underlying $image. Perhaps a bug in InterventionImage::createCloneWithResource?

**Comment by tractorcow** [comment]
Oh, DOH!

It's because InterventionImage is creating a temp file and that file is NOT segmented based on SHA1 or variant! This means multiple resizes will overwrite the same temp location.

**Comment by tractorcow** [comment]
I think it's this `$this->setTempPath($path);`

The temp path isn't set correctly for the cloned image resource.

**Comment by tractorcow** [comment]
Solution: Don't use a temp file.

**Comment by tractorcow** [comment]
Ah, seems to be more than just that... will keep digging.

**Comment by tractorcow** [comment]
ok, it seems to be a basic mutability issue with the backend itself.

```
return $this->createCloneWithResource($resource->crop($width, $height, $left, $top));
```

This code DOES clone the resource, but it clones it AFTER the operation is performed, meaning it's modifying the root object first.

createCloneWithResource should take a callback to be called AFTER the clone, not before it.

I'll fix tomorrow.

**Comment by tractorcow** [comment]
@dhensby do you want to look at this mutability issue before I get my grubby hands on it?

File: src/ImageManipulation.php [hunk 1]
```diff
@@ -581,20 +581,25 @@ public function getIcon()
...
      */
     public function getImageBackend()
     {
-        // Skip if non-image or file is broken
-        if (!$this->getIsImage() || !$this->exists()) {
-            return null;
+        // Pass to backend service factory
+        $backend = Injector::inst()->createWithArgs(Image_Backend::class, array($this));
+        if ($backend instanceof Image_Backend) {
+            return $backend;
```
Comment (flamerohr) [inline src/ImageManipulation.php]: Is this to ensure that `InvalidBackend` doesn't make it through?

Also noticed `$this->image_backend_cache` hasn't been used yet, and may not be needed anymore :)

File: src/ImageManipulation.php [hunk 2]
```diff
@@ -581,20 +581,25 @@ public function getIcon()
     }
 
     /**
+     * Cache of image backends
+     *
+     * @var array
+     */
+    protected static $image_backend_cache = [];
```
Comment (kinglozzer) [inline src/ImageManipulation.php]: This can be removed now, right?

File: src/ImageBackendFactory.php [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        $key = sha1($store->getHash().'-'.$store->getVariant());
+        if (isset($this->cache[$key])) {
+            return $this->cache[$key];
+        }
+
+        // Create appropriate result
+        if ($store->exists() && $store->getIsImage()) {
+            $backend = $this->creator->create($service, $params);
+        } else {
+            $backend = new InvalidBackend();
```
Comment (flamerohr) [inline src/ImageBackendFactory.php]: so I guess  `InvalidBackend` is just that dummy so cache doesn't invalidate, a little bit sly.

**Comment by kinglozzer** [comment]
One thing I’ve noticed in my testing (unrelated to this I think): when uploading a transparent png in assets, the copies that use `Fit` all end up with black backgrounds.

File: src/ImageManipulation.php (hunk 2 above)
Comment (tractorcow) [inline src/ImageManipulation.php]: Yeah that was a mistake. Leftover from a previous attempt.

Comment (sminnee) [inline src/ImageManipulation.php]: As an aside, should we deprecate snake case for non-config statics?

**Comment by tractorcow** [comment]
Ok, I've updated this PR with mutability protection. Tests are green!

If we can ever address https://github.com/Intervention/image/issues/745 we can also get rid of the temp file write; Another potential performance boost there. :)

File: src/ImageManipulation.php (hunk 2 above)
Comment (tractorcow) [inline src/ImageManipulation.php]: We should be using non-config statics extremely sparingly in any case. We should use psr-2.

File: src/ImageBackendFactory.php [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+     * @param string $service The class name of the service.
+     * @param array $params The constructor parameters.
+     * @return object The created service instances.
+     */
+    public function create($service, array $params = array())
+    {
+        /** @var AssetContainer */
+        $store = reset($params);
+        if (!$store instanceof AssetContainer) {
+            throw new BadMethodCallException("Cannot only create Image_Backend for " . AssetContainer::class);
```
Comment (tractorcow) [inline src/ImageBackendFactory.php]: Ok fixed now.

File: src/ImageBackendFactory.php (hunk 3 above)
Comment (tractorcow) [inline src/ImageBackendFactory.php]: I got rid of this class. Now I just catch the injector exception, but this should only happen if you have a broken image anyway (getIsImage() should prevent this being called at the top level).

File: src/ImageManipulation.php (hunk 1 above)
Comment (tractorcow) [inline src/ImageManipulation.php]: Fixed and removed the static.

**Comment by tractorcow** [comment]
@mikenz have you any feedback from your testing?

**Comment by sminnee** [comment]
I tested this PR with a folder of 185 x 1mb files and it's gone from timing out to running out of memory.

A snip of the call stack is here. It's reading the every file with Image_Backend to get the width/height metadata for the GraphQL responses. This needs to be persistently cached somehow (or saved to the DB)

Ideally, the cache for this data would be populated on upload.

```
96    8.0835    520243016    GraphQL\Executor\Executor::resolveField( )    .../Executor.php:227
97    8.0835    520243272    GraphQL\Executor\Executor::resolveOrError( )    .../Executor.php:442
98    8.0835    520243328    call_user_func:{/Users/sminnee/Sites/ss4/vendor/webonyx/graphql-php/src/Executor/Executor.php:488} ( )    .../Executor.php:488
99    8.0835    520243328    SilverStripe\GraphQL\TypeCreator->SilverStripe\GraphQL\{closure}( )    .../Executor.php:488
100    8.0835    520243704    call_user_func_array:{/Users/sminnee/Sites/ss4/graphql/src/TypeCreator.php:189} ( )    .../TypeCreator.php:189
101    8.0835    520244144    SilverStripe\AssetAdmin\GraphQL\FileTypeCreator->resolveField( )    .../TypeCreator.php:189
102    8.0835    520244144    SilverStripe\GraphQL\Util\CaseInsensitiveFieldAccessor->getValue( )    .../FileTypeCreator.php:248
103    8.0837    520244560    SilverStripe\View\ViewableData->__get( )    .../CaseInsensitiveFieldAccessor.php:66
104    8.0838    520244600    SilverStripe\Assets\File->getHeight( )    .../ViewableData.php:128
105    8.0845    520244840    SilverStripe\Assets\InterventionBackend->getHeight( )    .../ImageManipulation.php:631
106    8.0845    520244840    SilverStripe\Assets\InterventionBackend->getImageResource( )    .../InterventionBackend.php:313
107    8.0846    520244840    SilverStripe\Assets\InterventionBackend->loadFromContainer( )    .../InterventionBackend.php:214
108    8.0896    520245256    Intervention\Image\ImageManager->make( )    .../InterventionBackend.php:170
109    8.0896    520245504    Intervention\Image\AbstractDriver->init( )    .../ImageManager.php:50
110    8.0896    520245504    Intervention\Image\AbstractDecoder->init( )    .../AbstractDriver.php:64
111    8.0903    520245600    Intervention\Image\Gd\Decoder->initFromPath( )    .../AbstractDecoder.php:338
112    8.1281    532089528    Intervention\Image\Gd\Decoder->gdResourceToTruecolor( )    .../Decoder.php:68
113    8.1281    532089528    imagecreatetruecolor ( )    .../Decoder.php:136
```

**Comment by tractorcow** [comment]
@sminnee I'll look at adding caching to image dimensions. That's probably the best I can do without implementing a complete filesystem cache;

**Comment by sminnee** [comment]
:+1:

**Pull request merged**